### PR TITLE
[meta] nansum, nanmedian (and few minor clean-ups)

### PR DESCRIPTION
**Pull request by kshitij12345**
meta support for `nansum` and `nanmedian`

### CODE REVIEW

File: torch/_meta_registrations.py [hunk 1]
```diff
@@ -529,3 +525,56 @@ def meta_embedding_bag_forward_only(weight, indices, offsets, *args):
...
+    output_size = _get_reduction_output_size(input, (), False)
+    return input.new_empty(output_size)
+
+
+@torch.library.impl(meta_lib, "nanmedian.dim_values")
+@torch.library.impl(meta_lib, "nanmedian.dim")
+@out_wrapper_multi("values", "indices")
+def meta_nanmedian_dim(input, dim=-1, keepdim=False):
+    output_size = _get_reduction_output_size(input, (dim,), keepdim)
+    return input.new_empty(output_size), input.new_empty(output_size, dtype=torch.long)
```
Comment (anjali411) [inline torch/_meta_registrations.py:580]: nit - `new_empty` sets the `requires_grad` to `False` by default but we should propagate it from input right?

Comment (kshitij12345) [reply torch/_meta_registrations.py:580]: It does set it to `False` (see below). Interestingly, none of the registration pass `requires_grad` (and tests don't complain). Should we update all ops and update the test to verify if `requires_grad` is propagated?

```python
import torch
from torch._subclasses.fake_tensor import (
    FakeTensor,
    FakeTensorMode,
)

def _to_tensormeta(x):
    mode = torch._prims.utils.get_prim_fake_mode()
    if isinstance(x, torch.Tensor):
        out = FakeTensor.from_tensor(x, mode)
        return out
    return x

t = torch.randn((), requires_grad=True)
assert t.new_empty((2, 2)).requires_grad == t.requires_grad  # Fails 
```

File: torch/_meta_registrations.py [hunk 2]
```diff
@@ -529,3 +525,56 @@ def meta_embedding_bag_forward_only(weight, indices, offsets, *args):
...
+
+    if input.dtype.is_floating_point or input.dtype.is_complex:
+        return input.dtype
+    elif promote_int_to_long:
+        return torch.long
+
+    return input.dtype
+
+
+def _get_reduction_output_size(input, dims, keepdim):
```
Comment (anjali411) [inline torch/_meta_registrations.py:543]: can we use `compute_reduction_output_shape` (See `meta_var_mean_correction`)?

**Review by anjali411 — APPROVED** [review]

File: torch/_meta_registrations.py (hunk 1 above)
Comment (ezyang) [reply torch/_meta_registrations.py:580]: The meta functions live below autograd so it doesn't matter, you'll always be running below what is effectively `no_grad`

**Comment by ezyang** [comment]
It's a little unfortunate as you have already written the helpers, but for these probably would've been the same amount of work or even less to write decompositions instead, those already have helpers for dealing with reductions

File: torch/_meta_registrations.py (hunk 2 above)
Comment (kshitij12345) [reply torch/_meta_registrations.py]: Updated. Thanks!

File: torch/_meta_registrations.py [hunk 3]
```diff
@@ -529,3 +527,45 @@ def meta_embedding_bag_forward_only(weight, indices, offsets, *args):
     if offsets.device.type == "cpu":
         bag_size = offsets.new_empty(offsets.size())
     return output, offset2bag, bag_size, max_indices
+
+
+def _get_reduction_dtype(input, dtype, promote_int_to_long=True):
```
Comment (anjali411) [inline torch/_meta_registrations.py:532]: Is this similar to `reduction_dtypes`?

File: torch/_meta_registrations.py [hunk 4]
```diff
@@ -505,20 +502,21 @@ def is_fast_path(src, scale, output, padding_idx):
     return output, offset2bag, bag_size, max_indices
 
 
+@torch.library.impl(meta_lib, "diag")
 @torch.library.impl(meta_lib, "diag.out")
-def meta_diag_out(self, dim=0, *, out):
-    assert self.dim() in (1, 2), "matrix or a vector expected"
+@out_wrapper
+def meta_diag(self, dim=0):
```
Comment (ezyang) [inline torch/_meta_registrations.py:508]: dim here is misnamed, it's called diagonal in native_functions.yaml

File: torch/_meta_registrations.py [hunk 5]
```diff
@@ -505,20 +502,21 @@ def is_fast_path(src, scale, output, padding_idx):
...
+        sz = min(self.size(0), self.size(1) - dim)
     else:
-        if dim >= 0:
-            sz = min(self.size(0), self.size(1) - dim)
-        else:
-            sz = min(self.size(0) + dim, self.size(1))
-        torch._resize_output_(out, (sz,), self.device)
-        return out
+        sz = min(self.size(0) + dim, self.size(1))
+    return self.new_empty((sz,))
```
Comment (ezyang) [inline torch/_meta_registrations.py:519]: I'm perplexed why this isn't a decomp, the C++ implementation is

```
Tensor diag(const Tensor& self, int64_t offset) {
  auto ndim = self.dim();
  TORCH_CHECK(ndim == 1 || ndim == 2, "diag(): Supports 1D or 2D tensors. Got ", self.dim(), "D"); 
  if (ndim == 1) {
    return at::diag_embed(self, offset);
  } else {
    // We return a copy of the diagonal
    return at::diagonal_copy(self, offset);
  }
} 
```

Comment (kshitij12345) [reply torch/_meta_registrations.py:519]: Even, I am not sure now 😓. But the following PR took care of it

https://github.com/pytorch/pytorch/pull/87180/files#diff-6dffed1ade0ba3e887f9a4eafa3bfcec267ab2365b8adcb91bd391f49b3fd2e3L1022
